src/odemis/model/_dataflow.py

```diff
@@ -580,34 +615,37 @@ def run(self):
 
                 # receive data
                 if self._data in socks:
-                    # TODO: be more resilient if wrong data is received (can
-                    # block forever)
+                    # TODO: be more resilient if wrong data is received (can block forever)
                     array_format = self._data.recv_pyobj()
                     array_md = self._data.recv_pyobj()
                     array_buf = self._data.recv(copy=False)
                     # logging.debug("Received new DataArray over ZMQ for %s", self.uri)
                     # more fresh data already?
-                    if (self._data.getsockopt(zmq.EVENTS) & zmq.POLLIN and
-                        discarded < self.max_discard):
+                    if (discarded < max_discard
+                        and self._data.getsockopt(zmq.EVENTS) & zmq.POLLIN
+                       ):
                         discarded += 1
                         # logging.debug("Discarding object received as a newer one is available")
                         continue
+                    # Don't log here, because if we are discarding message it's because we are running
+                    # out of time, and logging is slow.
                     # TODO: only log the accumulated number every second, to avoid log flooding
-#                     if discarded:
-#                         logging.debug("Dataflow %s dropped %d arrays", self.uri, discarded)
+                    if discarded:
+                        logging.warning("Dataflow %s dropped %d arrays", self.uri, discarded)
                     discarded = 0
                     # TODO: any need to use zmq.utils.rebuffer.array_from_buffer()?
                     if len(array_buf):
                         array = numpy.frombuffer(array_buf, dtype=array_format["dtype"])
-                    else: # frombuffer doesn't support zero length array
+                    else:  # frombuffer doesn't support zero length array
                         array = numpy.empty((0,), dtype=array_format["dtype"])
                     array.shape = array_format["shape"]
                     darray = DataArray(array, metadata=array_md)
+                    self.weak_df.notify(darray)
 
-                    try:
-                        self.w_notifier(darray)
-                    except WeakRefLostError:
-                        return  # It's a sign there is nothing left to do
+        except ReferenceError:  # The DataFlow(Proxy) is gone
+            # => stop this thread too
+            logging.debug("Dataflow proxy %s is gone, stopping the subscription thread", self.uri)
+            return
         except Exception:
             if logging:
                 logging.exception("Ending ZMQ thread due to exception")
```

Comment:
does it make sense to log the exception here?

Reply:
Yes, it's quite useful, especially in a thread. Otherwise, any error will stop the thread execution, without *any* feedback.

Reply:
I forgot that `logging.exception` also logs the raised exception, so you can disregard my comment.